wapitiCore/attack/mod_csrf.py

```diff
@@ -0,0 +1,203 @@
+                return original_request.headers[header]
+
+        return None
+
+    def is_csrf_robust(self, csrf_value):
+        """Check whether anti-csrf token is robust enough"""
+
+        # csrf value length is too short
+        if len(csrf_value) < self.MIN_LENGTH:
+            return False
+
+        # csrf value entropy is too low, i.e. csrf value is too predictive
+        if self.entropy(csrf_value) < self.MIN_ENTROPY:
+            return False
+
+        return True
+
+    @staticmethod
+    def is_same_response(original_response: Page, mutated_response: Page):
+        """Check whether mutated_response status and content are identical to those of original_response"""
+
+        if original_response.status != mutated_response.status:
+            return False
+
+        if original_response.content != mutated_response.content: # TODO: Maybe too strict
+            return False
+
+        return True
+
+    def is_csrf_verified(self, original_request: Request):
+        """Check whether anti-csrf token is verified (backend) after submitting request"""
+
+        # Replace anti-csrf token value from form with "wapiti"
+        mutated_post_params = [
+            param if param[0] != self.csrf_string else [self.csrf_string, "wapiti"]
+            for param in original_request.post_params
+        ]
+
+        # Replace anti-csrf token value from headers with "wapiti"
+        mutated_headers = original_request.headers
+        if self.csrf_string in mutated_headers:
+            mutated_headers[self.csrf_string] = "wapiti"
+
+        mutated_request = Request(
+            path=original_request.path, method=original_request.method,
+            get_params=original_request.get_params, post_params=mutated_post_params,
+            file_params=original_request.file_params, referer=original_request.referer,
+            link_depth=original_request.link_depth
+        )
+
+        original_response = self.crawler.send(original_request, follow_redirects=True)
+        mutated_response = original_response
+
+        try:
+            mutated_response = self.crawler.send(mutated_request, headers=mutated_headers, follow_redirects=True)
+
+        except ReadTimeout:
+
+            self.log_orange("---")
+            self.log_orange(Anomaly.MSG_TIMEOUT, original_request.path)
+            self.log_orange(Anomaly.MSG_EVIL_REQUEST)
+            self.log_orange(mutated_request.http_repr())
+            self.log_orange("---")
+
+            anom_msg = Anomaly.MSG_PARAM_TIMEOUT.format(self.csrf_string)
+
+            self.add_anom(
+                request_id=original_request.path_id,
+                category=Anomaly.RES_CONSUMPTION,
+                level=Anomaly.MEDIUM_LEVEL,
+                request=mutated_request,
+                info=anom_msg,
+            )
+
+        else:
+            if mutated_response.status == 500:
+                anom_msg = Anomaly.MSG_PARAM_500.format(self.csrf_string)
+
+                self.add_anom(
```

Comment:
We don't necessary need to check for 500 error here. It is more useful when we are using an aggresive payload (like those in sql / exec / file modules)